### ISSUE TITLE
Lists cause generic-random to generate infinite values

### ISSUE THREAD

**Issue by feuerbach** [Lysxia/generic-random#12]
Despite using `genericArbitraryU'`, generic-random seems to generate infinite values when the type recursion happens inside a list.

Example:

``` haskell
{-# LANGUAGE DeriveGeneric, TypeApplications #-}

import GHC.Generics
import Control.Monad
import Generic.Random (genericArbitraryU')
import Test.QuickCheck

data Tree = Leaf | Node [Tree]
  deriving (Generic, Show)

instance Arbitrary Tree where
  arbitrary = genericArbitraryU'

main = replicateM_ 100 $ do
  print =<< generate (arbitrary @Tree)
```

**Comment by Lysxia** [comment]
Thanks for the report! I forgot to decrease the size when the constructor is unary.

**Comment by Lysxia** [comment]
Currently this generator still explodes in size because `arbitrary` for lists doesn't reduce the size parameter, so nodes close to the root have high arity.

Does that seem like a problem which is easy enough to run into that it would be a good default to replace the generator when a field is a list with one that divides the size by say, the length of the list?

**Comment by Lysxia** [comment]
Now fixed in `1.2.0.0`. I added a special case for lists, which is a crude default but avoids explosions in many simple cases. It can be disabled easily or replaced with something more fine-tuned using the custom generator mechanism described in my tutorial module.

**Comment by feuerbach** [comment]
Thanks!
